Exposed pipes will now leak their contents

BlueNexus opened the pull request:
<!-- 
Do not forget to add a changelog when you have made admin/player facing changes that can alter gameplay.
Examples which require a changelog entry include:
* Adding/removing objects that players may interact with, or the way they function.
* Adding/removing/altering admin tools.
* Changing the map.

Examples were changelog entries are optional/not typically required:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

You find a README and example file in .\html\changelogs\ for further instructions.
-->

**This is the second half of the pipe bursting PR**.

Simply put, any pipe which abruptly ends without something to close the loop (such as a connector, or an endcap (they're finally useful!)) will leak its contents into the air. **(Basically any pipe with the "exposed" sprite, shown below)**.

In addition, I have added pipe clamps, which can be placed on a pipe to halt the flow of gas inside that pipe. This is useful for isolating sections of the loop in the event of a breach, or shutting off a part of the loop so changes can be made to it.

Finally, Automated Shutoff Valves have been added and placed around the map. These will attempt to lessen the damage caused by a major leak, by closing when the pipe's pressure falls too low, and opening again when the leaking stops.

Example of a pipe which will leak:
![srhb1er](https://cloud.githubusercontent.com/assets/12956760/23340869/acb66a40-fc35-11e6-8881-f648848506e0.png)

The main purpose of this is so burst pipes will leak, but it has many other effects as well.

I would like to point out that pipes have very low volume when compared with a room. The room will not overpressurise instantly - Even in Atmospherics, with 15MPa pipes, the room maybe goes up 2kPa per second, and not forever.

Todo: 
- [x] Add a device for sealing off sections of pipe, for engineering work. and isolating parts of the loop in the event of a major leak. (Stasis Clamp)
- [x] Make Stasis Clamps easily accessible to engineers
- [x] Add Automated Shutoff Valves to control the damage done by a major leak
- [x] Add Automated Shutoff Valves to the map
- [x] Un-double-up the ES valves
- [x] Test!
- [x] Fix this - https://i.imgur.com/MfZfltZ.gifv

BlueNexus committed:
broken pipes now leak

Kasuobes commented:
Add corks and wedges made from wood. We subsim now.

atlantiscze commented:
I feel this particular change is pushing it a bit too far.

Few serious problems this will cause:
- Sometimes you need to temporarily disconnect a piece of machinery that's connected to pipes (such as a freezer when you are upgrading it), and having to wait for pipeline to depressurise is just a waste of time
- The moment we have a breach somewhere with damaged piping, whole distro will leak out through there. This is probably the most severe problem.
- It makes various tasks, such as hot-swapping pumps for high powered ones on engine impossible, as well as making even marginal damage to piping extremely dangerous both for atmospherics and the engine itself.

So yeah, :-1: on this for me. I kind of didn't mind the first PR, but this is /way/ too far into "annoying behavior" zone.

BlueNexus commented:
@atlantiscze All of your points are dealt with by the thing in my to-do list. The cryogenic clamp will allow you to shut off portions of the loop so you can work on them. You can also use this to shut off pipes leading to leaky areas, so you don't lose all your air.

Also, these clamps will be very readily available.

atlantiscze commented:
Except for the fact that you'll need minutes (at least) as an engineer to get to the area with broken pipes. By that time, damage has already been done, and a ton of air was lost. Still fully against this as it needlessly overcomplicates a system, to which everyone's been used to for a long time, without adding any real benefits.

BlueNexus commented:
Air doesn't leak out /that/ quickly. In the time it takes for an engineer to reach the area, it won't have done any major damage that atmospherics can't fix. I'd like to point out that atmos is /more/ than capable of restoring any damage done by a leak. 

This is a necessary part of pipes bursting, and a little complexity is fine.

atlantiscze commented:
I'm mostly talking about damage this causes /to atmospherics/

Amount of gas in atmospherics pressure tanks is very limited, especially on the torch. I can imagine that with this change, even few breaches which involve damaged piping leaking air would very quickly result in loss of large amounts of gas.

In general, atmospherics, the piping layout, and even the game aren't ready for this. In real life applications there would probably be large amount of computer controlled valves that would automatically activate in the event of pipe breach, limiting the leak to minimal to none levels. Some modern airplanes have this for hydraulic system, for example. It cuts off the damaged section, allowing the rest to operate.

Sounds familiar? This is what we have now, pretty much. Pipes being destroyed by explosions already leak a little bit of gas (notable for high pressure pipelines, like engine pipes), but they don't suffer from more leaks.

BlueNexus commented:
Atmospherics works fine. I've tested it extensively - It never reaches the bursting threshold. Atmos is capped at 15MPa.

If you would like me to add automated clamps to control damage done during a breach, I will happily do so. It wouldn't be that hard to do.

atlantiscze commented:
You are changing the topic of this discussion. I never talked about bursting pressures. I am talking about how easy it is for atmospherics to run out of air with leaks caused due to pipe damage with these changes you are proposing.

BlueNexus commented:
I suggest a test, then. How would you like me to prove that the torch's atmos can handle some leakage? Maybe let it leak for a while and see if it can recover? @atlantiscze

BlueNexus committed:
Adds cryogenic clamps

atlantiscze commented:
I'll only note that atmospherics substations wouldn't resolve anything, as you would still need someone to operate them and cut off the gas flow once a pipe damage occurs.

The only viable solution would be fully automatic clamps, but they'd have to be very frequent and fully automated, as cutting off whole deck due to one breach is a no-no. And, if you think about it a bit, we actually have this - pipes already do leak when damaged, but only their contents leak. Everything else does, and should stay inside the pipe.

This PR is only overcomplicating already complex enough system. While i don't frequently play on other servers, it's been like this pretty much everywhere where i played on. Introducing sudden leaks from open pipes is going to put many players into situations where they accidentally cause a gas leak. And don't get me even started on talking about what could happen to the engine when a single pipe on the core loop is broken off...

BlueNexus commented:
Causing a major leak accidentally can be easily avoided via the use of clamps. I'll look into automated clamps - I also still think the system can handle it. As I said earlier, I propose a test, maybe on a test server, to see if the system can recover from a leak.

atlantiscze commented:
We can arrange something later today, contact me on IRC

BlueNexus commented:
Asked on discord. People seem to be generally supportive of this - Definitely going to try and get automatic clamps in place, though.

BlueNexus committed:
Renamed cryogenic clamps

They are now named Stasis Clamps.

Ithalan commented:
As I mentioned on discord, I like the direction that this PR is taking atmos handling in, but that is pre-conditioned on the atmos system being updated with safety shut-off valves that automatically isolate sections of the system in which a breach is detected, as well as portable clams to isolate a section of piping for being worked on.

Down the road I'd like to see the atmos layout fully redesigned to incorporate substations for each deck or department and ideally a second distro pipe that can either be used to supply custom gas mixes to connector ports in the substations, or used by malicious people to feed the vents in a particular department

BlueNexus commented:
@Ithalan Portable clamps are already finished, I just need to work on the automatic systems.

BlueNexus commented:
Turns out that I can just use the existing pressure regulator pipes rather than adding automated clamps. This will be finished today.

atlantiscze commented:
Pressure regulators have diametrally different function than automatic clamps would.

PREGs can be used as passive gate that aims for specific pressure on either input or output. Mapping these around the ship won't fix anything, as they don't automatically close when one pipeline is breached, and will still let gas through. If anything, they will make things far more annoying for engineers, as everyone prefers to set the distribution loop to different pressure.

So, yeah. Still against this as whole.

BlueNexus commented:
I see. I'll make the automated clamps after all, then.

BlueNexus commented:
@raptor1628 has agreed to help out with adding atmospheric substations after this PR has been merged.

atlantiscze commented:
"Let's first merge a PR that will make a single breach threaten the whole ship's supply, and then consider possible fixes to that problem"

BlueNexus commented:
You misunderstand. I'm still going to put in the automated clamps before this is merged, the substations are a later thing.

BlueNexus committed:
Added Emergency Shutoff Valves

BlueNexus commented:
Automated clamps are done. Placing them on the map.

Techhead0 commented on code/modules/atmospherics/components/shutoff.dm:
```diff
@@ -0,0 +1,39 @@
+	var/node2_pressure = node2_air.return_pressure()
+	if(node1_last_pressure && node2_last_pressure)
+		if(open)
+			if(((node1_pressure <= threshold) || (node2_pressure <= threshold)) && (!((node1_pressure >= node1_last_pressure) && (node2_pressure >= node2_last_pressure)))) //I'm not proud of this line. /BlueNexus
```
This messy statement is logically equivalent to:
```
if(((node1_pressure <= threshold) || (node2_pressure <= threshold))\
&& ((node1_pressure < node1_last_pressure) || (node2_pressure < node2_last_pressure)))
```
ie. "Pressure is below threshold in one side and pressure is not rising in either."
However, I think the following might be smarter:
```
if(((node1_pressure <= threshold) && (node1_pressure < node1_last_pressure)\
|| ((node2_pressure <= threshold) && (node2_pressure < node2_last_pressure)))
```
ie. "Pressure in one side is below threshold and not rising."

(I also used `\` to break up the line and make it easier to read. Lining up the two halves of the second example is a coincidence.)

BlueNexus replied:
Will look into this once everything else is finished

BlueNexus replied:
Not going to use the second version, since it causes roundstart problems while the loop initially fills up. Will revisit the first one tomorrow.

Techhead0 replied:
On further reading, if the system is open, `node1_pressure` should always equal `node2_pressure`. I'll also come back to this in a bit.

BlueNexus committed:
Maps in ES Valves

Also renamed emergency shutoff valves to automated shutoff valves

BlueNexus committed:
Final tweaks

BlueNexus commented on maps/torch/torch-3.dmm:
```diff
@@ -580,8 +580,6 @@
 "lj" = (/obj/effect/floor_decal/industrial/warning{icon_state = "warning"; dir = 8},/turf/simulated/floor/plating,/area/engineering/engine_room)
-"lk" = (/obj/machinery/atmospherics/pipe/simple/hidden/supply{dir = 10},/turf/simulated/floor/plating,/area/maintenance/seconddeck/aftport)
-"ll" = (/obj/machinery/atmospherics/pipe/simple/hidden/scrubbers,/obj/structure/disposalpipe/segment,/turf/simulated/floor/plating,/area/maintenance/seconddeck/aftport)
 "lm" = (/turf/simulated/wall/r_wall,/area/vacant/cannon)
```
Did I mess up here?

Techhead0 replied:
Those two items look like they got sorted to the end.
`lk`->`Fh` and `ll`->`Fi`.

BlueNexus replied:
Alright, perfect.

BlueNexus commented:
This is ready for review.

mustafakalash commented:
Let's see a forum vote, please.

BlueNexus commented:
I'll make it when I get home.
https://baystation12.net/forums/threads/vote-broken-pipes-now-leak-new-valves.3764/
@mustafakalash

BlueNexus commented:
Going to fix a few things while I wait on the vote.

BlueNexus commented:
Can someone tell me how to revert that merge commit and try again?

BlueNexus commented:
I'm going to need some help resolving these conflicts.

BlueNexus committed:
fixes

BlueNexus committed:
Clamp icon fix

BlueNexus committed:
fixes and tweaks

BlueNexus commented:
Testing complete, everything works fine.

BlueNexus commented:
![xkppuc3](https://cloud.githubusercontent.com/assets/12956760/23500881/8dddd33c-ff29-11e6-942d-f494d2664bf5.png)
Poll is overwhelmingly in favour @mustafakalash

BlueNexus committed:
Clamps now attach to the targeted pipe

BlueNexus commented:
> _Day four of the pipe leaking PR. Upon investigating runtime 52, I discovered that the clamps had been nonfunctional the entire time. I turn to build_pipeline() for a solution - The archaic code confuses me; It seems alien in its design, using concepts I cannot begin to understand - Ones that should not function at all. I continue looking for a solution amongst the ancient code._

_The following is scratched below, hastily written_

> _I see victory in the distance. By mistake, I have been given insight into the problem, and a hint to the solution. My spirits are redoubled once more._

> Today I rejoice. The Solution is mine!

BlueNexus committed:
Fixes clamps

BlueNexus commented:
This is ready for review, and good to go.

BlueNexus commented:
Waiting on @PsiOmegaDelta (Who for some reason volunteered to handle the conflicts via merge commit)

BlueNexus commented:
Fixing travis

BlueNexus commented:
Travis clearly hates me.

Techhead0 commented:
IIRC, Travis can't do tests if you have merge conflicts at the time of commit.

BlueNexus commented:
Welp, that didn't work

BlueNexus commented:
Aaaaaaa
How do I make this merge work ;-;
`Fatal: Git write-tree failed to write a tree`

BlueNexus committed:
Conflict resolution part 1

BlueNexus commented:
This isn't ready, don't merge yet

BlueNexus committed:
Merge branch 'dev' into pipeleaks

BlueNexus committed:
conflict resolution 2

BlueNexus committed:
conflict resolution 3

BlueNexus commented:
Okay, this should finally be ready.

BlueNexus commented:
Okay, this should finally be ready for review.

Poll is here
https://baystation12.net/forums/threads/vote-broken-pipes-now-leak-new-valves.3764/

BlueNexus committed:
Adds ASV to robotics

BlueNexus committed:
AS valves can now be reset

Should make it easy to deal with valve malfunctions, by forcing it to reconsider.

Techhead0 commented on code/game/machinery/atmoalter/clamp.dm:
```diff
@@ -0,0 +1,139 @@
+
+/obj/machinery/clamp/New(loc, var/obj/machinery/atmospherics/pipe/simple/to_attach = null)
+	..()
+	if(to_attach)
```
Would be safer as `istype(to_attach)`.

BlueNexus replied:
Resolved.

Techhead0 commented on code/modules/atmospherics/datum_pipeline.dm:
```diff
@@ -79,6 +79,9 @@ datum/pipeline
 								air.merge(item.air_temporary)
 
+						else if(item.in_stasis)
+							edge_check++
```
Put this check first and you shouldn't have to check `item.in_stasis` more than once.

Also, instead of incrementing and decrementing `edge_check`, consider just using a `continue` to skip to the next pipe.

BlueNexus replied:
Resolved.

Techhead0 commented on code/modules/atmospherics/datum_pipeline.dm:
```diff
@@ -99,9 +102,14 @@ datum/pipeline
+					if(!istype(result,/obj/machinery/atmospherics/pipe) && (result!=reference) && !P.in_stasis)
+						result.network_expand(new_network, edge)
+				else if(!istype(result,/obj/machinery/atmospherics/pipe) && (result!=reference))
 					result.network_expand(new_network, edge)
```
This entire for loop is a huge mess. A bunch of checks added, some redundant, to make something equivalent to what it was before.

BlueNexus replied:
How would I accomplish this without these checks? I considered putting the in_stasis var under machinery/atmospherics.

Techhead0 replied:
How do you accomplish what? Your added stuff doesn't accomplish anything.
Look at the if statements and istypes carefully.

BlueNexus replied:
Resolved, this line just wasn't changed

Techhead0 commented on code/game/machinery/atmoalter/clamp.dm:
```diff
@@ -0,0 +1,139 @@
+	if(to_attach)
+		target = to_attach
+	else
+		target = locate(/obj/machinery/atmospherics/pipe/simple) in loc
```
Educational sidenote: This line is equivalent to `target = locate() in loc` since `locate()` will assume the path if none is explicitly supplied to it, like `istype()`.

No need to change it, though.

BlueNexus committed:
Resolves comments

BlueNexus committed:
Tweaks and visual changes

BlueNexus commented:
I can't squash this due to git errors, sadly.

Here's what the floor marker looks like

![](https://i.imgur.com/p1ZxDc6.png)

BlueNexus committed:
Removes unnecessary typechecks

Techhead0 reviewed (approved):
Tentative approval. The thing I'm most hesitant about is the clamp's open and close procs.

BlueNexus committed:
Fixes a runtime error

BlueNexus committed:
Resolves conflicts

BlueNexus committed:
Merge branch 'dev' into pipeleaks

BlueNexus commented:
Mwerezak hasn't been active in a month, and SkyMarshal doesn't know Pipecode.
Merge please, I'm fed up of resolving conflicts by redoing my changes from scratch.

BlueNexus committed:
Corrects a map error

Techhead0 commented on maps/torch/torch-4.dmm:
```diff
@@ -1,4 +1,4 @@
-"aa" = (/turf/space,/area/space)
+"aa" = (/turf/space/transit/east,/area/space)
```
Why the hell was this changed?

BlueNexus replied:
I have no idea, that wasn't something I changed explicitly. I'll fix it.

BlueNexus replied:
Resolved.

BlueNexus committed:
Fixes

BlueNexus commented:
Now just to hope that no map PRs are merged before this is 😢
